src/core/FaceControls.tsx

```diff
@@ -0,0 +1,413 @@
+          }
+
+          explCamera.position.copy(current.position)
+          explCamera.rotation.copy(current.rotation)
+        }
+      },
+      [explCamera, computeTarget, smoothTime, current.position, current.rotation]
+    )
+
+    //
+    // detect()
+    //
+
+    const [faces, setFaces] = useState<FaceLandmarkerResult>()
+    const faceLandmarker = useFaceLandmarker()
+    const detect = useCallback<FaceControlsApi['detect']>(
+      (video, time) => {
+        const faces = faceLandmarker?.detectForVideo(video, time)
+        setFaces(faces)
+      },
+      [faceLandmarker]
+    )
+
+    useFrame((_, delta) => {
+      if (!manualUpdate) {
+        update(delta)
+      }
+    })
+
+    // Ref API
+    const api = useMemo<FaceControlsApi>(
+      () =>
+        Object.assign(Object.create(THREE.EventDispatcher.prototype), {
+          detect,
+          computeTarget,
+          update,
+          facemeshApiRef,
+          webcamApiRef,
+          // shorthands
+          play: () => {
+            webcamApiRef.current?.videoTextureApiRef.current?.texture.source.data.play()
+          },
+          pause: () => {
+            webcamApiRef.current?.videoTextureApiRef.current?.texture.source.data.pause()
+          },
+        }),
+      [detect, computeTarget, update]
+    )
+    useImperativeHandle(fref, () => api, [api])
+
+    //
+    // events callbacks
+    //
+
+    useEffect(() => {
+      const onVideoFrameCb = (e: THREE.Event) => {
+        if (!manualDetect) detect(e.texture.source.data, e.time)
+        if (onVideoFrame) onVideoFrame(e)
+      }
+
+      api.addEventListener('videoFrame', onVideoFrameCb)
+
+      return () => {
+        api.removeEventListener('videoFrame', onVideoFrameCb)
+      }
+    }, [api, detect, faceLandmarker, manualDetect, onVideoFrame])
+
+    // `controls` global state
+    useEffect(() => {
+      if (makeDefault) {
+        const old = get().controls
+        set({ controls: api })
+        return () => set({ controls: old })
+      }
+    }, [makeDefault, api, get, set])
+
+    const points = faces?.faceLandmarks[0]
+    const facialTransformationMatrix = faces?.facialTransformationMatrixes?.[0]
+    const faceBlendshapes = faces?.faceBlendshapes?.[0]
+    return (
+      <FaceControlsContext.Provider value={api}>
+        {webcam && <Webcam ref={webcamApiRef} autostart={autostart} videoTextureSrc={webcamVideoTextureSrc} />}
+
+        <Facemesh
+          ref={facemeshApiRef}
+          {...facemesh}
+          points={points}
+          depth={depth}
+          facialTransformationMatrix={facialTransformationMatrix}
+          faceBlendshapes={faceBlendshapes}
+          eyes={eyes}
+          eyesAsOrigin={eyesAsOrigin}
+          offset={offset}
+          offsetScalar={offsetScalar}
+          debug={debug}
+          rotation-z={Math.PI}
+          visible={debug}
+        >
+          <meshStandardMaterial flatShading={true} side={THREE.DoubleSide} />
```

Comment:
Could a `MeshBasicMaterial` also suffice or would this need to react to lighting. I'd consider a `MeshPhongMaterial` instead otherwise for performance. three's `MeshStandardMaterial` is much too expensive for mobile devices or large screen coverage.

Reply:
done: converted to `MeshBasicMaterial`